src/main/java/br/com/deveficiente/bolaoapi/services/poll/api/InvitationEmailController.java

```diff
@@ -20,19 +27,58 @@
 public class InvitationEmailController {
 
     private final InvitationRepository invitationRepository;
+    private final UserRepository userRepository;
+    private final PollRepository pollRepository;
 
-    public InvitationEmailController(InvitationRepository invitationRepository) {
+    public InvitationEmailController(InvitationRepository invitationRepository, UserRepository userRepository, PollRepository pollRepository) {
         this.invitationRepository = invitationRepository;
+        this.userRepository = userRepository;
+        this.pollRepository = pollRepository;
     }
 
     @Transactional
     @GetMapping
-    public ResponseEntity process(@RequestParam String key) {
+    public ResponseEntity process(@RequestParam String key, @RequestParam Boolean accept, UriComponentsBuilder uriBuilder) {
         Optional<Invitation> invitation = invitationRepository.findByKeyAndExpirationAfter(key, LocalDateTime.now());
 
-        if (!invitation.isPresent())
-            return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new ErrorResponse("invitation link expired."));
+        //check invitation
+        if (!invitation.isPresent()) return invalidInvitation();
 
-        return ResponseEntity.ok(new InvitationAccepted(invitation.get()));
+        //check acceptance
+        if (!accept) return declineInvitation(invitation);
+
+        //check user
+        String invitedUserLogin = invitation.get().getEmail();
+        Optional<User> invitedUser = userRepository.findByLogin(invitedUserLogin);
+
+        if (!invitedUser.isPresent()) return userNotExists(uriBuilder);
+
+        return ResponseEntity.ok(acceptInvitation(invitation, invitedUser));
+    }
+
+    private ResponseEntity invalidInvitation() {
+        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(new ErrorResponse("invitation link not exists or expired."));
+    }
+
+    private ResponseEntity userNotExists(UriComponentsBuilder uriBuilder) {
+        URI uri = uriBuilder.path("/users").build().toUri();
+        HttpHeaders headers = new HttpHeaders();
+        headers.setLocation(uri);
+        return ResponseEntity.status(HttpStatus.SEE_OTHER).headers(headers).build();
+    }
+
+    private ResponseEntity declineInvitation(Optional<Invitation> invitation) {
+        invitation.get().decline();
+        this.invitationRepository.save(invitation.get());
```

Comment:
*Usamos tudo que conhecemos que está pronto. Só fazemos código do zero se for estritamente necessário.*

Não precisa dessa linha aqui. Como a invitation é gerenciada, você não precisa chamar o save. O hibernate vai perceber a diferenca no estado final do objeto e vai gerar o update automaticamente. Dado que você marcou o método com transactional.